### ISSUE TITLE
Rule idea: must-use-await [code included]

### ISSUE THREAD

**Issue by mikemaccana** [eslint/eslint#8830]
I'm currently refactoring code to remove functions named `cb` in favour of `await` (and `util.promisify()` where necessary).

I suspect I'm not alone here.

It would be useful to have a rule to find cases where a function has a final argument names `cb` or `callback` and flag it.

**Comment by eslintbot** [comment]
Hi @mikemaccana, thanks for the issue. It looks like there's not enough information for us to know how to help you.

If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

Requesting a rule change? Please see [Proposing a Rule Change](http://eslint.org/docs/developer-guide/contributing/rule-changes) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by mikemaccana** [comment]
@platinumazure I've read http://eslint.org/docs/developer-guide/contributing/new-rules - I had hoped to able to suggest a rule that was useful to many people and find someone in eslint who would champion it. What specifically is missing?

**Comment by onurtemizkan** [comment]
Hi @mikemaccana, Your rule proposal should be in [this](https://github.com/eslint/eslint/blob/master/templates/rule-proposal.md) format.

**Comment by mikemaccana** [comment]
Note: I codes this as a plugin today - it's at https://www.npmjs.com/package/eslint-plugin-must-use-await, but I still think it would be better in eslint itself.

**Please describe what the rule should do:**



**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[ ] Warns about a potential error
[X] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
function foo(yolo, swag, cb){}

function foo(banana, grape, done){}

function foo(callback){}

foo.then(function(){})
```

**Why should this rule be included in ESLint (instead of a plugin)?**

Since ES2017 many JS coding in [direct style](https://en.wikipedia.org/wiki/Direct_style) has become quite popular. It's a new part of the language itself, and a lot of people are refactoring old code to use it - it would be great is eslint could help.

**Comment by not-an-aardvark** [comment]
Could you clarify what the rule would actually do? I'm confused about what your examples have in common.

**Comment by mikemaccana** [comment]
@not-an-aardvark Flag examples of code where await should be used instead.

![deel6ruxsaeuows](https://user-images.githubusercontent.com/172594/27922808-35501d0c-6274-11e7-872a-8c95be9f9935.jpg)

**Comment by mikemaccana** [comment]
I've renamed this feature request to match the name of the existing plugin (which is also more descriptive)

**Comment by peet** [comment]
I believe what you are looking for also exists in (eslint-plugin-promise)[https://github.com/xjamundx/eslint-plugin-promise], `specifically prefer-await-to-callbacks` and `prefer-await-to-then`

**Comment by not-an-aardvark** [comment]
Thanks for clarifying.

Personally, I'm not in favor of adding this to core -- I think there are a lot of cases where it's better to use `then`. However, let's see what the rest of the team thinks.

**Comment by mikemaccana** [comment]
Ack @peet - I suspect it's a fairly common need. Hence asking for it in core!

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get enough support from the team and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after a long time tend to never do it, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.

**Comment by mikemaccana** [comment]
Thanks @not-an-aardvark for being polite and I understand completely. Have a good one and thanks for making eslint!
